File: DSA/not_yet_category/1338_Reduce_Array_Size_to_The_Half/Reduce_Array_Size_to_The_Half.py
"""
输入：arr = [3,3,3,3,5,5,5,2,2,7]
输出：2
解释：选择 {3,7} 使得结果数组为 [5,5,5,2,2]、长度为 5（原数组长度的一半）。
大小为 2 的可行集合有 {3,5},{3,2},{5,2}。
选择 {2,7} 是不可行的，它的结果数组为 [3,3,3,3,5,5,5]，新数组长度大于原数组的二分之一。
"""
import time

#explan 1
import collections
class Solution:
    def minSetSize(self, arr):
        #dict計算各有幾個
        cnt = collections.Counter(arr)
        print(1, cnt)
        f = list(cnt.values())
        print(2, f)
        #value小到大
        f.sort()
        print(3, f)

        ans, removed, half = 0, 0, len(arr) // 2
        print(4, ans,removed,half)
        while removed < half:
            ans += 1
            print(5, ans)
            #從最大開始+ 超過原長度一半即停止
            removed += f.pop()
            print(6, removed)
        return ans

c = [5,5,1,3,4,4,5,2]
test = Solution()

start = time.process_time()
print(test.minSetSize(c))
end = time.process_time()
print("執行時間：%f 秒" % (end-start))

File: DSA/not_yet_category/1338_Reduce_Array_Size_to_The_Half/Reduce_Array_Size_to_The_Half.go
package main
import (
	// "fmt"
    "sort"
)

func minSetSize(arr []int) int {
	countDict := make(map[int]int)
    maxCount := 0
    for _, val := range arr {
        countDict[val]++ //maintain a map based on val : count of this val
        maxCount = max(maxCount, countDict[val]) //track the max count
    }
    //get value & sort slice
    bucket := make([]int, 0, len(countDict))
    for _, val := range countDict {
        bucket = append(bucket, val)
    }
    sort.Ints(bucket)

    res := 0
    removed := 0
    half := len(arr)/2                  //總長度的一半
    for removed < half {
        res += 1    
        removed += bucket[len(bucket)-1] //加上最後一個元素
        bucket = bucket[:len(bucket)-1]  //刪除最後一個元素
    }
    // fmt.Println(res)
	return res
}

func max(a, b int) int {
    if a > b {
        return a
    } else {
        return b
    }
}

func main(){
	arr := []int{5,5,1,3,4,4,5,2}
	minSetSize(arr)
}
    // sort.Ints(bucket)
    // fmt.Println(bucket)
    // keys := make([]int, 0, len(countDict))
    // for k := range countDict {
	// 	keys = append(keys, k)
	// 	fmt.Println(keys)
    // }

    // for key, value := range countDict {
    //     fmt.Println("key:", key, "value::",value)
    // }
    